server/app/controllers/TaskSolver.scala

```diff
@@ -58,32 +57,46 @@ class TaskSolver @Inject()(executor: RuntimeSuiteExecutor with DynamicSuiteExecu
     }.recover { case NonFatal(e) => notFound }
   }
 
-  def taskStream: WebSocket = WebSocket.accept { req =>
-    val channel: PublishChannel[Event] = PublishChannel[Event](DropOld(20))
-    val sink = Sink.foreach { clientInput: JsValue =>
-      val prevTimestamp = (clientInput \ "prevTimestamp").as[Long]
-      val currentTimestamp = (clientInput \ "currentTimestamp").as[Long]
-      if (Duration(currentTimestamp - prevTimestamp, TimeUnit.MILLISECONDS) < 1.seconds) {
-        channel.pushNext(Line("Too many requests per second from the same client. Slow down"))
-        channel.pushComplete()
-      } else {
-        val solution = (clientInput \ "solution").as[String]
-        val year = (clientInput \ "year").as[Long]
-        val lang = (clientInput \ "lang").as[String]
-        val timeuuid = (clientInput \ "timeuuid").as[String]
-
-        getCachedTask(year, lang, UUID.fromString(timeuuid)).onComplete {
-          case Success(Some(task)) =>
-            channel.pushNext(service.testResult(Try(executor(solution, task.suite, task.solutionTrait)(s => channel.pushNext(Line(s))))))
-            channel.pushComplete()
-          case Success(None) =>
-            channel.endWithError(new RuntimeException(s"Task is not available for a given solution: $solution"))
-          case Failure(ex) =>
-            channel.endWithError(ex)
+  def taskStream: WebSocket = WebSocket.accept { _ =>
+    val clientInputPromise = Promise[JsValue]()
+    val channel: Observable[Event] =
+      Observable.create[Event](OverflowStrategy.DropOld(20)) { downstream =>
+        clientInputPromise.future.onComplete {
+          case Success(clientInput) =>
+            val prevTimestamp = (clientInput \ "prevTimestamp").as[Long]
+            val currentTimestamp = (clientInput \ "currentTimestamp").as[Long]
+            if (Duration(currentTimestamp - prevTimestamp, TimeUnit.MILLISECONDS) < 1.seconds) {
+              downstream.onNext(Line("Too many requests per second from the same client. Slow down"))
+              downstream.onComplete()
+            } else {
+              val solution = (clientInput \ "solution").as[String]
+              val year = (clientInput \ "year").as[Long]
+              val lang = (clientInput \ "lang").as[String]
+              val timeuuid = (clientInput \ "timeuuid").as[String]
+
+              getCachedTask(year, lang, UUID.fromString(timeuuid)).onComplete {
+                case Success(Some(task)) =>
+                  val block: (String => Unit) => String = executor(solution, task.suite, task.solutionTrait)
+                  val testResult: Try[String] => TestResult = service.testResult
```

Comment:
This place is carbon copy of `service.ObservableRunner#apply`, same code structure will help to rewrite it later